operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go

```diff
@@ -0,0 +1,349 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"flag"
+	"fmt"
+	"os"
+	"path/filepath"
+	"reflect"
+	"runtime"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/concurrency"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework/command"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+
+	"github.com/grafana/mimir/pkg/mimir"
+)
+
+const MimirImage = "grafana/mimir"
+
+type ValueAnnotator struct {
+	Value string `yaml:"value" json:"value"`
+}
+
+func main() {
+	config := new(ValueAnnotator)
+	fn := func(items []*yaml.RNode) ([]*yaml.RNode, error) {
+		mapper := NewConfigExtractor(items)
+		configs, err := mapper.ResolveConfigs()
+		if err != nil {
+			return nil, err
+		}
+		return append(items, configs...), nil
+	}
+	p := framework.SimpleProcessor{Config: config, Filter: kio.FilterFunc(fn)}
+	cmd := command.Build(p, command.StandaloneDisabled, false)
+	command.AddGenerateDockerfile(cmd)
+	if err := cmd.Execute(); err != nil {
+		os.Exit(1)
+	}
+}
+
+type ConfigExtractor struct {
+	allItems []*yaml.RNode
+}
+
+func NewConfigExtractor(items []*yaml.RNode) *ConfigExtractor {
+	return &ConfigExtractor{
+		allItems: items,
+	}
+}
+
+// PodSpec is a subset from k8s.io/api/core/v1.PodSpec
+// copied here in order to avoid a dependency on the k8s.io/api/core/v1 package
+type PodSpec struct {
+	Volumes []struct {
+		Name      string `yaml:"name"`
+		ConfigMap *struct {
+			Name string `yaml:"name"`
+		} `yaml:"configMap"`
+		Secret *struct {
+			SecretName string `yaml:"secretName"`
+		} `yaml:"secret"`
+		EmptyDir *struct{}
+	} `yaml:"volumes"`
+	Containers []struct {
+		Args         []string `yaml:"args"`
+		Image        string   `yaml:"image"`
+		VolumeMounts []struct {
+			Name      string `yaml:"name"`
+			MountPath string `yaml:"mountPath"`
+		} `yaml:"volumeMounts"`
+	} `yaml:"containers"`
+}
+
+// DeploymentOrStatefulSet is a subset from k8s.io/api/apps/v1.Deployment and k8s.io/api/apps/v1.StatefulSet
+// copied here in order to avoid a dependency on the k8s.io/api/apps/v1 package
+type DeploymentOrStatefulSet struct {
+	Spec struct {
+		Template struct {
+			Spec PodSpec `yaml:"spec"`
+		} `yaml:"template"`
+	} `yaml:"spec"`
+}
+
+func (c *ConfigExtractor) ExtractPodSpec(obj *yaml.RNode) (PodSpec, bool, error) {
```

Comment:
Any reason for keeping this a method vs a regular function?

Reply:
nope, updated